graalpy/graalpy-micronaut-pygal-charts/README.md

```diff
@@ -0,0 +1,45 @@
+## SVG Charts with GraalPy and Micronaut
+
+This demo illustrates how GraalPy can be used to embed [Pygal](https://github.com/Kozea/pygal), a dynamic SVG charting library written in Python, in a Micronaut application.
+In particular, this demo shows four different approaches to interact with Pygal from Java.
+
+## Preparation
+
+Install GraalVM for JDK 23 and set the value of `JAVA_HOME` accordingly.
+We recommend using [SDKMAN!](https://sdkman.io/). (For other download options, see [GraalVM Downloads](https://www.graalvm.org/downloads/).)
+
+```bash
+sdk install java 23-graal
+```
+
+## Run the Application
+
+To start the demo, simply run:
+
+```bash
+./mvnw package mn:run
+```
+
+When the demo runs, open the follwing URLs in a browser:
+
+| URL                           | Service                       |
+|:------------------------------|:------------------------------|
+| http://localhost:8080/java    | [`PyGalServicePureJava`](src/main/java/com/example/PyGalServicePureJava.java)        |
+| http://localhost:8080/python  | [`PyGalServicePurePython`](src/main/java/com/example/PyGalServicePurePython.java)      |
+| http://localhost:8080/mixed   | [`PyGalServiceMixed`](src/main/java/com/example/PyGalServiceMixed.java)           |
+| http://localhost:8080/dynamic | [`PyGalServiceValueAPIDynamic`](src/main/java/com/example/PyGalServiceValueAPIDynamic.java) |
+
+
+## Implementation Details
+
+The `DemoController` uses four services that all render the same XY chart using different implementations:
+
+- [`PyGalServicePureJava`](src/main/java/com/example/PyGalServicePureJava.java) interacts with Pygal and Python using Java interfaces and `Value.as(Class<T> targetType)`. This is the recommended approach.
+- [`PyGalServicePurePython`](src/main/java/com/example/PyGalServicePurePython.java) embeds the Python sample code from the Pygal documentation.
+- [`PyGalServiceMixed`](src/main/java/com/example/PyGalServiceMixed.java) uses a Python function which is invoked with Java values.
+- [`PyGalServiceValueAPIDynamic`](src/main/java/com/example/PyGalServiceValueAPIDynamic.java) uses the [Value](https://www.graalvm.org/sdk/javadoc/org/graalvm/polyglot/Value.html) API from the GraalVM SDK.
+
+
+The `DemoTest` ensures that all four service implementations render the same XY chart. Run it with `./mvnw test`.
+
+> Note: This demo uses a single [`GraalPyContext`](src/main/java/com/example/GraalPyContext.java), which can be accessed from only one thread at a time to evaluate Python code. Pure Python packages including Pygal can be used in multiple GraalPy contexts, for example one context per thread, to improve the throughput of the application. Other demos illustrate how to pool multiple contexts.
```

Comment:
I find the phrasing "accessed from only one thread at a time" potentially misleading. JavaScript Context cannot be accessed from multiple threads at all - it should throw runtime exception AFAIK. Python Context can be, but under the hood it will serialize the execution. Some ideas how to phrase this:

"...which can execute Python code only in one thread at a time." (short, leaves out details)

"...which can execute Python code only in one thread at a time for compatibility reasons."

"...which can execute Python code only in one thread at a time. Threads running Python code are internally scheduled in round-robin fashion." (to avoid the possible impression that one Python thread can starve all others)

Reply:
Maybe with a link to https://docs.python.org/3/glossary.html#term-global-interpreter-lock

Reply:
done!